### PR TITLE
Initial GPU Support

### PR DESCRIPTION
**Pull request by majst01**
Dependend PRs:

- [x] https://github.com/metal-stack/metalctl/pull/240/
- [x] https://github.com/metal-stack/metal-go/pull/158
- [x] https://github.com/metal-stack/metal-hammer/pull/121
- [x] https://github.com/metal-stack/metal-roles/pull/274
- [x] https://github.com/u-root/u-root/pull/2950
- [ ] https://github.com/metal-stack/docs/pull/192

We need to test a scenario where multiple pods need to access the same GPU:

There are several documents from nvidia on this topic

https://developer.nvidia.com/blog/improving-gpu-utilization-in-kubernetes
https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-operator-mig.html
https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-sharing.html

### CODE REVIEW

File: cmd/metal-api/internal/metal/size.go [hunk 1]
```diff
@@ -84,6 +87,31 @@ func (c *Constraint) Matches(hw MachineHardware) (ConstraintMatchingLog, bool) {
...
+	case GPUConstraint:
+		existing := make(map[string]uint8)
+		for _, gpu := range hw.MetalGPUs {
+			_, ok := existing[gpu.Model]
+			if !ok {
+				existing[gpu.Model] = 1
+			} else {
+				existing[gpu.Model]++
+			}
+		}
```
Comment (ulrichSchreiner) [inline cmd/metal-api/internal/metal/size.go:99]: i think the code

~~~go
existing[gpu.Model] = existing[gpu.Model] +1 
~~~

should also work.

File: cmd/metal-api/internal/metal/size.go [hunk 2]
```diff
@@ -126,7 +154,7 @@ func (s *Size) overlaps(so *Size) bool {
 	}
 	for _, c := range s.Constraints {
 		for _, co := range so.Constraints {
-			if c.Type == co.Type && ((c.Min < co.Min && c.Max < co.Min) || (c.Min > co.Min && c.Min > co.Max)) {
+			if c.Type == co.Type && ((c.Min < co.Min && c.Max < co.Min) || (c.Min > co.Min && c.Min > co.Max) || !reflect.DeepEqual(c.GPUs, co.GPUs)) {
 				return false
```
Comment (ulrichSchreiner) [inline cmd/metal-api/internal/metal/size.go:158]: why not `maps.Equal`?

Comment (majst01) [reply cmd/metal-api/internal/metal/size.go:158]: done

File: cmd/metal-api/internal/metal/size.go (hunk 1 above)
Comment (majst01) [reply cmd/metal-api/internal/metal/size.go:99]: But why ?

Comment (ulrichSchreiner) [reply cmd/metal-api/internal/metal/size.go:99]: less verbose, but it's ok do to it this way.

File: cmd/metal-api/internal/service/v1/machine.go [hunk 3]
```diff
@@ -83,15 +83,28 @@ type MachineNetwork struct {
...
 
 type MachineHardwareBase struct {
-	Memory   uint64               `json:"memory" description:"the total memory of the machine"`
-	CPUCores int                  `json:"cpu_cores" description:"the number of cpu cores"`
-	Disks    []MachineBlockDevice `json:"disks" description:"the list of block devices of this machine"`
+	Memory    uint64               `json:"memory" description:"the total memory of the machine"`
+	CPUCores  int                  `json:"cpu_cores" description:"the number of cpu cores"`
+	Disks     []MachineBlockDevice `json:"disks" description:"the list of block devices of this machine"`
+	MetalCPUs []MetalCPU           `json:"cpus" description:"the cpu details"`
+	MetalGPUs []MetalGPU           `json:"gpus" description:"the gpu details"`
```
[the remark this answers is not in the document]
Comment (majst01) [reply cmd/metal-api/internal/service/v1/machine.go:90]: done

File: cmd/metal-api/internal/metal/size.go [hunk 4]
```diff
@@ -84,6 +88,24 @@ func (c *Constraint) Matches(hw MachineHardware) (ConstraintMatchingLog, bool) {
...
+			if res {
+				break
+			}
+		}
+
+		if c.Identifier == "" {
+			res = false
+		}
+
+		cml.Log = fmt.Sprintf("existing gpus:%#v required gpus:%s count %d-%d", hw.MetalCPUs, c.Identifier, c.Min, c.Max)
```
[the remark this answers is not in the document]
Comment (majst01) [reply cmd/metal-api/internal/metal/size.go:108]: thanks, done

File: cmd/metal-api/internal/metal/size.go [hunk 5]
```diff
@@ -84,6 +88,24 @@ func (c *Constraint) Matches(hw MachineHardware) (ConstraintMatchingLog, bool) {
...
+			}
+			res = count >= c.Min && count <= c.Max && idMatches
+			if res {
+				break
+			}
+		}
+
+		if c.Identifier == "" {
+			res = false
+		}
```
[the remark this answers is not in the document]
Comment (majst01) [reply cmd/metal-api/internal/metal/size.go:106]: validated against required identifier for gpu constraints

File: cmd/metal-api/internal/metal/size_test.go [hunk 6]
```diff
@@ -79,6 +79,124 @@ var (
...
+			},
+			{
+				Type:       GPUConstraint,
+				Max:        4,
+				Min:        4,
+				Identifier: "H100*",
+			},
+		},
+	}
+	// mixedGPUSize = Size{
```
[the remark this answers is not in the document]
Comment (majst01) [reply cmd/metal-api/internal/metal/size_test.go:166]: Is already checked in TestSize_Validate#two constraints with same type

**Review by Gerrit91 — APPROVED** [review]

**Pull request merged**